bokeh/util/future.py

```diff
@@ -56,3 +58,57 @@ def wraps(obj, attr_names=WRAPPER_ASSIGNMENTS, wraps=wraps):
         return wraps(obj, assigned=(name for name in attr_names if hasattr(obj, name)))
 
 del partial, WRAPPER_ASSIGNMENTS
+
+
+# inspect.getargspec and inspect.formatargspec were deprecated in Python 3.5
+# in favor of the newer inspect.signature introspection
+
+if sys.version_info[:2] < (3, 4):
+
+    def signature(func):
+        # The modifications in this function are to make results more in line
+        # with Python 3, i.e. self is not included in bound methods, supplied
+        # parameters are not reported in partial, etc. This simplifies the
+        # downstream code considerably.
+        from inspect import getargspec, isfunction, ismethod
+        from functools import partial
+
+        if isfunction(func) or ismethod(func):
+            sig = getargspec(func)
+            if ismethod(func):
+                sig.args.remove('self')
+            return sig
+
+        elif isinstance(func, partial):
+            sig = getargspec(func.func)
+            if 'self' in sig.args: sig.args.remove('self')
+            for name in func.keywords.keys():
+                sig.args.remove(name)
+            for val in func.args:
+                del sig.args[0]
+            return sig
+
+        else:
+            sig = getargspec(func.__call__)
+            sig.args.remove('self')
+            return sig
+
+    def format_signature(sig):
+        from inspect import formatargspec
+        return formatargspec(*sig)
+
+    def get_param_info(sig):
+        return (sig.args, sig.defaults or [])
```

Comment:
Not super happy with the modification to `args` here but they seem to work fine (all tests pass), this code is only used when there are errors anyway, and some day maybe we can drop python 2 support and delete all of this.